### PR TITLE
Make multiple calls example

### PR DESCRIPTION
**Pull request by atis06**

### CODE REVIEW

File: examples/caller/make_multiple_calls.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+input_context_parameters = {
+    k: context_pb2.Context.Parameter(
+        display_name=k,
+        value=v,
+        value_original=v,
+    )
+    for k, v in input_context_parameters.items()
+}
+
+call_id: str = "not_a_cai_session"  # change this
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py]: use str(uuid.uuid4())

File: examples/caller/make_multiple_calls.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        phone_number=phone_number,
+        call_id=call_id,
+        sip_sim_version=sip_sim_version,
+        project_id=PROJECT_ID,
+    )
+    return response
+
+
+##### Start multi-threaded calls #####
+threads: List[Thread] = [
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py:41]: nice :+1:

File: examples/caller/make_multiple_calls.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+CAI_CLIENT_CONFIG: str = "cert/path.json"
+#     PROJECT
+PROJECT_ID: str = "example_project_id"
+INIT_TEXT: str = "hello"
+
+##### Configurate #####
+
+PHONE_NUMBERS = ["+11233232", "+2342345"]
+
+sip_sim_version = SIP_SIM_VERSION
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py]: just use SIP_SIM_VERSION?

File: examples/caller/make_multiple_calls.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    AudioConfiguration,
+    CaiConfiguration,
+    VtsiConfiguration,
+)
+from ondewo.vtsi.voip_pb2 import StartCallInstanceResponse
+
+##### Globals #####
+
+#     SIP
+SIP_SIM_VERSION: str = "1.5.4"
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py:10]: can you just use the 'get_minimal_client' function, and make another script called 'configure_vtsi_server.py' where you show off all this config stuff?

I would like to end up with a few scripts, each demonstrating a simple idea-- this one should just have the multiple calls material in it

File: examples/caller/make_multiple_calls.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    s2t_type=S2T_TYPE,
+    s2t_language=S2T_LANGUAGE,
+)
+config_asterisk = AsteriskConfiguration(
+    host=ASTERISK_HOST,
+)
+
+##### Set contexts #####
+
+input_context_parameters = {
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py]: put this in a seperate file called 'configure_context_parameters.py'

**Comment by FergusFettes** [comment]
before you act on the comments in the file-- can you just rename the file 'full_demo.py', and make a new one called 'make_multiple_calls.py'.

because in the file you have made, you are demoing many different things-- how to do a full config, how to set contexts, and how to make multiple calls. please make seperate files for all these things.

also write loads of comments! these files should take a total newbie by the hand

File: examples/caller/make_multiple_calls.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+call_id: str = "not_a_cai_session"  # change this
+context = context_pb2.Context(
+    name=f"projects/{PROJECT_ID}/agent/sessions/{call_id}/contexts/input",
+    lifespan_count=10000,
+    lifespan_time=10000,
+    parameters=input_context_parameters,
+)
+
+contexts = [context]
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py]: for the contexts demo file, use this stuff i made for lukas:
```
#########################
#########################
# Helper Functions
#########################
#########################


def create_parameter_dict(my_dict: Dict) -> Optional[Dict[str, context_pb2.Context.Parameter]]:
    assert isinstance(my_dict, dict) or my_dict is None, "parameter must be a dict or None"
    if my_dict is not None:
        return {
            key: context_pb2.Context.Parameter(
                display_name=key,
                value=my_dict[key]
            )
            for key in my_dict
        }
    return None


def append_context(params, name):
    context = context_pb2.Context(
        name=f"projects/{project_id}/agent/sessions/{context_session_id}/contexts/{input_context_name}",
        lifespan_count=10000,
        parameters=create_parameter_dict(input_context_parameters),
    )
    contexts.append(context)

input_context_parameters = {
    "Timezone": "Monday from 2-4pm",
    "Supplier": "Amazon",
    "Object": "Waschmaschine",
    "Adresse": "Alliertenstraße 14/14, 1020 Wien",
    "Name": "Fergus Fettes",
    "Installation": "",
    "Payment": "",
}
input_context_name = 'c-config'
append_context(input_context_parameters, input_context_name)

input_context_parameters = {
    "Installation": "",
    "Payment": "",
}
input_context_name = 'c-config2'
append_context(input_context_parameters, input_context_name)
```

Comment (FergusFettes) [reply examples/caller/make_multiple_calls.py]: feel free to edit it of course

File: examples/caller/make_multiple_calls.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+##### Start call #####
+
+config_cai = CaiConfiguration(
+    host=CAI_HOST,
+    port=CAI_PORT,
+    cai_project_id=PROJECT_ID,
+    cai_contexts=contexts,
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py]: you can either add contexts here, or you can add them per call. for the contexts demo file, please note both methods. this means you can have the same contexts for all calls, or you can have different contexts for each call (if the context contains the name of the callee for instance)

**Review by FergusFettes — CHANGES_REQUESTED** [review]

File: examples/caller/configure_vtsi_server.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    cai_contexts=[],
+)
+
+# ConfigManager to use VTSI client and reach the configs of the different modules
+manager = ConfigManager(
+    config_audio=config_audio_ondewo,
+    config_cai=config_cai,
+    config_voip=config_voip,
+    config_asterisk=config_asterisk,
+)
```
Comment (FergusFettes) [inline examples/caller/configure_vtsi_server.py]: i would end all the examples with starting a basic listener

File: examples/caller/configure_vtsi_server.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+# CAI/BPI configuration
+# I. Input contexts can be passed here (passed contexts will be used for every calls)
+# or in VtsiClient.start_caller() call
+config_cai = CaiConfiguration(
+    host=CAI_HOST,
+    port=CAI_PORT,
+    cai_project_id=PROJECT_ID,
+    cai_contexts=[],
```
Comment (FergusFettes) [inline examples/caller/configure_vtsi_server.py]: delete this line

File: examples/caller/configure_vtsi_server.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+CAI_CLIENT_CONFIG: str = "cert/path.json"
+#     PROJECT
+PROJECT_ID: str = "example_project_id"
+
+########################################
+# Full configuration for VTSI, Asterisk, SIP, T2S, S2T modules
+
+# VTSI configuration
+config_voip = VtsiConfiguration(
+    host=VTSI_HOST, port=VTSI_PORT, secure=VTSI_SECURE, cert_path=VTSI_CERT
```
Comment (FergusFettes) [inline examples/caller/configure_vtsi_server.py]: make multiline

File: examples/caller/configure_context_parameters.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    name=f"projects/{PROJECT_ID}/agent/sessions/{call_id}/contexts/{input_context_name}",
+    lifespan_count=10000,
+    parameters=create_parameter_dict(input_context_parameters),
+)
+
+# Input context list (several context can be added)
+contexts_to_pass = [context]
+
+# Context list can be passed during CaiConfiguration initialization (global) OR
+# VtsiClient.start_caller() function call (local)
```
Comment (FergusFettes) [inline examples/caller/configure_context_parameters.py:65]: use my function for appending contexts!

also end it with a call.

show both the contexts added to the cai config and the contexts added to the call

File: examples/caller/make_multiple_calls.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+def deploy_caller(phone_number: str) -> StartCallInstanceResponse:
+    # Input contexts can be passed here (passed contexts will be used for only THIS call) or
+    # in CaiConfig initialization
+    response: StartCallInstanceResponse = client.start_caller(
+        init_text=INIT_TEXT,
+        phone_number=phone_number,
+        call_id=str(uuid.uuid4()),
+        sip_sim_version=SIP_SIM_VERSION,
+        project_id=PROJECT_ID,
+        # contexts=[]
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py]: delete!

File: examples/caller/make_multiple_calls.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+PHONE_NUMBERS = ["+11233232", "+2342345"]
+
+# get_minimal_client() returns the minimal working VTSI client
+client: VtsiClient = VtsiClient.get_minimal_client(voip_host=VTSI_HOST, voip_port=str(VTSI_PORT),
+                                                   secure=VTSI_SECURE, cert_path=VTSI_CERT)
+
+
+# deploy_caller() function for multi-threading
+def deploy_caller(phone_number: str) -> StartCallInstanceResponse:
+    # Input contexts can be passed here (passed contexts will be used for only THIS call) or
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py]: move this comment to the contexts example

File: examples/caller/make_multiple_calls.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+#     VTSI_SERVER
+VTSI_HOST: str = "0.0.0.0"
+VTSI_PORT: int = 12345
+VTSI_SECURE: bool = False  # if true, VTSI cert is needed
+VTSI_CERT: str = ""
+
+#     PROJECT
+PROJECT_ID: str = "example_project_id"
+INIT_TEXT: str = "hello"
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py]: dont need init text, delete

Comment (FergusFettes) [reply examples/caller/make_multiple_calls.py]: make a comment in the config file about the init text, and what it does

File: examples/caller/make_multiple_calls.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+from typing import List
+
+from ondewo.vtsi.client import VtsiClient
+from ondewo.vtsi.voip_pb2 import StartCallInstanceResponse
+
+#     SIP
+SIP_SIM_VERSION: str = "1.5.4"
+
+#     VTSI_SERVER
+VTSI_HOST: str = "0.0.0.0"
```
Comment (FergusFettes) [inline examples/caller/make_multiple_calls.py]: delete all this, you dont need it-- the defaults will use grpc-vtsi.ondewo.com. so the only thing you need is the grpc certificate. make a note of that.

File: examples/caller/configure_vtsi_server.py (hunk 8 above)
Comment (FergusFettes) [reply examples/caller/configure_vtsi_server.py]: so each of the example scripts should work to make a call if you run them. some calls will have different configs, some calls will have contexts, some calls will have multiple targets etc etc

**Review by FergusFettes — CHANGES_REQUESTED** [review]

File: examples/caller/make_multiple_calls.py (hunk 15 above)
Comment (atis06) [reply examples/caller/make_multiple_calls.py]: get_minimal_client() requires VTSI_HOST and VTSI_PORT, I can delete them only in the case of full client config

**Review by FergusFettes — APPROVED** [review]

**Pull request merged**